### PR TITLE
chore: [codex] map fused warmup and recover CUDA OOM fallback

### PR DESCRIPTION
**Pull request by ooples**
## Summary
- Expose `LinearWarmupScheduler` parameters needed by the compiled/fused optimizer bridge.
- Map `LinearWarmupScheduler` to `AiDotNet.Tensors` fused `LrSchedule.LinearWarmup(...)`, preserving constant, linear-decay, and cosine-decay modes.
- Pin the local capture package pair used for the HarmonicEngine proof run.

## Verification
- `dotnet test tests\AiDotNet.Tests\AiDotNetTests.csproj --framework net10.0 --filter FullyQualifiedName~FusedLrScheduleMappingTests` passed 7/7.
- HarmonicEngine fused local-package canary passed with `HE_ALLOW_FUSED=1`: `avg_loss 9.4550 -> 7.4261`, `pL1 40621.47 -> 46411.97`, no eager fallback.

## Summary by CodeRabbit

* **New Features**
  * Added support for faster GPU-backed inference and training paths in supported models.
  * Exposed additional learning-rate schedule details for easier inspection.

* **Bug Fixes**
  * Improved handling of GPU memory and transient failures during training.
  * Refined profiling so timing better reflects actual device work.
  * Better fallback behavior when optimized GPU paths aren’t available.

* **Tests**
  * Added integration coverage for learning-rate schedule matching and GPU failure classification.

### CODE REVIEW

**Comment by ooples** [comment]
Update after the HRE Cortex audit:

- Added committed fused-plan CUDA OOM classification for driver exceptions like `cuMemAllocAsync failed: Out of memory`.
- Retires compiled fused state, drops dead GPU activation cache, preserves sticky fused-disable after reset, and forces single-pass streaming clipping for OOM recovery.
- Added targeted classification coverage and reused the existing fast-clip convergence coverage.

Validation:
- `dotnet test tests\AiDotNet.Tests\AiDotNetTests.csproj --framework net10.0 --no-restore --filter FullyQualifiedName~FusedOptimizerIntegrationTests.CudaAllocationOom_IsClassifiedAsStreamingOom|FullyQualifiedName~FusedOptimizerIntegrationTests.FastClip_SinglePass_ReducesLoss_AndStaysFinite -p:UseSharedCompilation=false -nr:false`
- Packed local `AiDotNet.0.227.8-capture.nupkg` and proved HRE `d256/L4/b1024` completes after the fused OOM boundary with moving loss/params.

**Comment by ooples** [comment]
Closing as **superseded by `master`**.

The core of this PR — the AiDotNet.Tensors **#642 deferred GPU execution graph** for diffusion plus the CodeRabbit **#1650** review fixes — has already landed on `master` through a separate path:
- `DiffusionModelBase.PredictNoise` has the deferred-recording scope, `PredictorStreamCapture`, and the `DiffusionModelOptions.UseGpuExecutionGraph` gate (its comments even cite "CodeRabbit #671/#1650").
- Both `DiffusionGpuExecutionGraphCudaTests.cs` and `DiffusionGpuExecutionGraphFallbackTests.cs` exist on `master` — a rebase here produces **add/add** conflicts on those exact files.

This branch is **64 commits behind** `master`, so the remaining CodeRabbit comments (restore-training-mode, force-CPU-engine-for-fallback, don't-disable-the-CUDA-gate, etc.) apply to a test/impl version that `master` has already replaced.

The two commits with genuinely-unique work not yet on `master` are tracked for separate salvage in **#1751**:
- `963343088` — fused fast-path mapping for `LinearWarmupScheduler` (blocked on publishing `LrSchedule.LinearWarmup`/`WarmupDecayMode` in Tensors; `master`'s 0.105.0 only has `LinearWarmupCosine`).
- `4909081c0` — committed-fused-CUDA-OOM → streaming recovery in `NeuralNetworkBase` (needs CUDA-HW verification + reconciliation with master's existing OOM handling).

No code is lost — #1751 preserves the salvageable pieces.

**Comment by ooples** [comment]
Codex update 2026-07-01:

Added commit `153b773e` for post-step training residency cleanup found while validating the HarmonicEngine Cortex run.

Root cause: `TrainWithTape` needed to keep full gradient semantics, but retained gradient maps and activation-cache state longer than necessary after optimizer diagnostics and stepping.

Fix:
- Keep `ComputeGradients(lossForBackward, sources: null)` so all trainable parameters still receive gradients.
- Clear returned gradient maps after diagnostics/step.
- Drop the GPU activation cache after the step to reduce post-training residency pressure.

Validation:
- net10 focused filters: 33/33 passed.
- net471 focused filters: 33/33 passed.
- Packed local `AiDotNet.0.227.10-capture.nupkg` and reran the HRE proof against it plus the local tensor package.

File: src/Diffusion/NoisePredictors/UNetNoisePredictor.cs [hunk 1]
```diff
@@ -485,24 +485,88 @@ private void CreateDefaultDecoderBlocks()
 
 
     /// <inheritdoc />
+    // #1650/#638 diffusion CUDA-graph capture (opt-in AIDOTNET_DIFFUSION_CUDA_GRAPH=1):
+    // captures the eager ForwardUNet into a CUDA graph once (warmup) and replays it per
+    // denoising step — collapsing the UNet's hundreds of kernel launches into one
+    // cuGraphLaunch. Per-instance; lazily created on first GPU forward.
+    private AiDotNet.Tensors.Engines.Gpu.ResidentInferenceGraph? _residentGraph;
+    private static readonly bool s_diffusionCudaGraph =
+        System.Environment.GetEnvironmentVariable("AIDOTNET_DIFFUSION_CUDA_GRAPH") == "1";
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Diffusion/NoisePredictors/UNetNoisePredictor.cs:497]: Fixed in ab5c3e0be: UNetNoisePredictor now reads AIDOTNET_DIFFUSION_CUDA_GRAPH via a live expression-bodied property instead of caching it at type-load, matching DiffusionModelBase.PredictorStreamCapture, so an env var set after startup is honored. Resolving.

[the remark this answers is not in the document]
Comment (ooples) [reply src/Diffusion/NoisePredictors/UNetNoisePredictor.cs]: Fixed in ab5c3e0be: UNetNoisePredictor now reads AIDOTNET_DIFFUSION_CUDA_GRAPH via a live expression-bodied property instead of caching it at type-load, matching DiffusionModelBase.PredictorStreamCapture, so an env var set after startup is honored. Resolving.

File: tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/FusedOptimizerIntegrationTests.cs [hunk 2]
```diff
@@ -96,6 +96,15 @@ public async Task FusedAdam_TrainingCompletes_WithFiniteLossAcrossSteps()
...
     }
 
+    [Fact]
+    public void CudaAllocationOom_IsClassifiedAsStreamingOom()
+    {
+        var ex = new InvalidOperationException("cuMemAllocAsync failed: Out of memory");
+
+        Assert.True(FusedTrainingTestNetwork.IsGpuOutOfMemoryForTest(ex));
+        Assert.True(FusedTrainingTestNetwork.IsGpuTransientForTest(ex));
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/FusedOptimizerIntegrationTests.cs:110]: Addressed in ab5c3e0be: added a comment clarifying the OOM string matches both classifiers by design, which is why recovery routing consults the OOM check first (the ordering in TryTrainWithFusedOptimizer is what makes it streaming-OOM recovery rather than a transient retry). Resolving.

File: Directory.Packages.props [hunk 3]
```diff
@@ -192,11 +192,12 @@
...
-         (AiDotNet.Tensors #680, merged + released). net8.0 in src/AiDotNet.csproj requires a
-         Tensors build with a net8.0 asset; the published 0.102.14 nupkg contains lib/net8.0/
-         (alongside net10.0 + net471), so this is no longer blocked. -->
-    <PackageVersion Include="AiDotNet.Tensors" Version="0.102.14" />
+    <!-- 0.104.8-capture (local-nuget) is the capture-branch (PR #705) Tensors that HRE runs: diffusion
+         CUDA-graph capture + all-backend FP16 conv/activations (#671) PLUS the at-scale capture-learns fixes
+         and the new LrSchedule.LinearWarmup mapping (warmup-then-Constant/Linear/Cosine on the fused captured
+         path — was falling to the eager tape, a 14-23x slowdown). Public path: 0.102.17 was the last published
+         release; the LinearWarmup + capture fixes go public via #705 then AiDotNet re-pins to that release. -->
+    <PackageVersion Include="AiDotNet.Tensors" Version="0.104.8-capture" />
```
[the remark this answers is not in the document]
Comment (ooples) [reply Directory.Packages.props]: Outdated: the referenced code has changed since this comment. This PR is held pending its Tensors dependency (capture build) publishing to nuget, so CI cannot restore/build it yet; any still-relevant version of this will resurface on re-review once it is buildable. Resolving the stale thread.

File: src/Diffusion/NoisePredictors/UNetNoisePredictor.cs [hunk 4]
```diff
@@ -485,24 +485,88 @@ private void CreateDefaultDecoderBlocks()
...
+            output = PredictCompiledForward(noisySample, timeEmbed, conditioning);
+        }
 
         _lastOutput = output;
         return streaming.Complete(output);
+
+        }
+        finally
+        {
+            if (evalForInference) SetAllLayersTrainingMode(true);
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Diffusion/NoisePredictors/UNetNoisePredictor.cs:558]: Outdated: the referenced code has changed since this comment. This PR is held pending its Tensors dependency (capture build) publishing to nuget, so CI cannot restore/build it yet; any still-relevant version of this will resurface on re-review once it is buildable. Resolving the stale thread.

File: tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphCudaTests.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        DirectGpuTensorEngine? gpu = null;
+        try { gpu = new DirectGpuTensorEngine(); }
+        catch { /* no CUDA backend */ }
+        Skip.If(gpu is null || !gpu.SupportsGpu || !gpu.IsGpuAvailable,
+            "No CUDA DirectGpuTensorEngine available on this machine.");
+        Skip.IfNot(TensorsDeferredGraphFixed,
+            "Blocked on a Tensors deferred-execution-graph (#642/#652) correctness bug: the deferred " +
+            "DDPM UNet denoise diverges ~100% from the eager-GPU forward (proven on a GTX 1660 Ti). " +
+            "Set AIDOTNET_TEST_DEFERRED_GRAPH=1 to run this once consuming a fixed Tensors build.");
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphCudaTests.cs]: Outdated: the referenced code has changed since this comment. This PR is held pending its Tensors dependency (capture build) publishing to nuget, so CI cannot restore/build it yet; any still-relevant version of this will resurface on re-review once it is buildable. Resolving the stale thread.

File: tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphFallbackTests.cs [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+
+        // Two models identical except for the flag; same seed -> same UNet weights + same
+        // initial noise + same scheduler trajectory. On the CPU engine the flag is inert
+        // (PredictNoiseStep sees no CUDA engine and calls PredictNoise directly), so the
+        // outputs must be element-wise identical — proving the fallback is transparent.
+        var eager = new DDPMModel<float>(architecture: null, options: Options(useGpuExecutionGraph: false), seed: seed);
+        var withFlag = new DDPMModel<float>(architecture: null, options: Options(useGpuExecutionGraph: true), seed: seed);
+
+        var eagerOutput = eager.Generate(shape, numInferenceSteps: steps, seed: seed);
+        var flagOutput = withFlag.Generate(shape, numInferenceSteps: steps, seed: seed);
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphFallbackTests.cs:58]: Outdated: the referenced code has changed since this comment. This PR is held pending its Tensors dependency (capture build) publishing to nuget, so CI cannot restore/build it yet; any still-relevant version of this will resurface on re-review once it is buildable. Resolving the stale thread.

File: tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphFallbackTests.cs [hunk 7]
```diff
@@ -0,0 +1,731 @@
...
+                Log($"DirectGpu ctor: SupportsGpu={ge.SupportsGpu} Name={ge.Name}");
+                if (ge.SupportsGpu)
+                {
+                    AiDotNet.Tensors.Engines.AiDotNetEngine.Current = ge;
+                    isCuda = true;
+                }
+            }
+            catch (System.Exception ex) { Log($"DirectGpu ctor FAILED: {ex.GetType().Name}: {ex.Message}"); }
+            if (!isCuda)
+                return; // CPU/CI: skip — the deferred GPU path only exists on CUDA hardware.
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphFallbackTests.cs:120]: Outdated: the referenced code has changed since this comment. This PR is held pending its Tensors dependency (capture build) publishing to nuget, so CI cannot restore/build it yet; any still-relevant version of this will resurface on re-review once it is buildable. Resolving the stale thread.

File: tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphFallbackTests.cs [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+                var key = System.Text.RegularExpressions.Regex.Replace(section, @"\[\d+\]", "");
+                var cur = totals.TryGetValue(key, out var v) ? v : (0.0, 0);
+                totals[key] = (cur.Item1 + ms, cur.Item2 + 1);
+            }
+            var sb = new System.Text.StringBuilder();
+            double grand = 0; foreach (var kv in totals) grand += kv.Value.ms;
+            foreach (var kv in System.Linq.Enumerable.OrderByDescending(totals, k => k.Value.ms))
+                sb.AppendLine($"{kv.Key,-24} {kv.Value.ms,9:F2} ms  {100 * kv.Value.ms / grand,5:F1}%  (n={kv.Value.n})");
+            sb.AppendLine($"{"TOTAL",-24} {grand,9:F2} ms");
+            try { System.IO.File.WriteAllText(System.IO.Path.Combine(System.IO.Path.GetTempPath(), "aidotnet_section_profile.txt"), sb.ToString()); } catch { }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/IntegrationTests/Diffusion/DiffusionGpuExecutionGraphFallbackTests.cs:238]: Outdated: the referenced code has changed since this comment. This PR is held pending its Tensors dependency (capture build) publishing to nuget, so CI cannot restore/build it yet; any still-relevant version of this will resurface on re-review once it is buildable. Resolving the stale thread.

File: src/NeuralNetworks/NeuralNetworkBase.cs [hunk 9]
```diff
@@ -7437,6 +7437,10 @@ void UnscaleAndCheck(Tensor<T>? g)
                     }
                 }
             }
+
+            allGrads.Clear();
+            grads.Clear();
+            GpuEngine?.DropActivationCache();
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/NeuralNetworks/NeuralNetworkBase.cs]: Valid perf concern (unconditional per-step GpuEngine.DropActivationCache). Not changing it in this PR: it is shared GPU-training code and this PR is currently held/unbuildable (its capture Tensors dependency is not published to nuget), so I cannot build or GPU-validate that gating the drop does not reintroduce the stale-cache issue the unconditional drop guards against. Leaving this open to fix + validate once the PR is buildable (Tensors #715 published).
